### PR TITLE
compute LDL*x

### PR DESCRIPTION
**Pull request by geoffroyleconte**
This may be useful when performing iterative refinement. It adds methods to the lmul! and mul! functions to compute LDL*x and LDL*X where x is a vector and X a matrix.

### CODE REVIEW

**Comment by dpo** [comment]
Thank you! Could you please add a test that checks that `ldl_mul!()` returns the same result as `A * x` when there is no regularization? Ideally, it should be a case with permutation.

File: test/runtests.jl [hunk 1]
```diff
@@ -230,3 +230,49 @@ end
...
+  b = A * ones(10)
+  T = eltype(A)
+  ϵ = sqrt(eps(T))
+  S = ldl_analyze(Symmetric(triu(A), :U))
+  ldl_factorize!(Symmetric(triu(A), :U), S)
+  x = copy(b)
+  ldiv!(S, x)
+  r1 = copy(x)
+  lmul!(S, r1)
+  @test norm(r1 - A * x) ≤ sqrt(ϵ) * norm(b)
```
Comment (dpo) [inline test/runtests.jl:255]:
```suggestion
  @test norm(r1 - A * x) ≤ sqrt(ϵ) * norm(r1)
```

File: test/runtests.jl [hunk 2]
```diff
@@ -230,3 +230,49 @@ end
...
+  r1 .-= b  # r1 = LDL*x-b
+  @test norm(r1) ≤ sqrt(ϵ) * norm(b)
+  mul!(r1, S, x) # same test using mul!
+  @test norm(r1 - A * x) ≤ sqrt(ϵ) * norm(b)
```
Comment (dpo) [inline test/runtests.jl:259]:
```suggestion
  @test norm(r1 - A * x) ≤ sqrt(ϵ) * norm(r1)
```

File: test/runtests.jl [hunk 3]
```diff
@@ -230,3 +230,49 @@ end
...
+  r1 .-= b
+  @test norm(r1) ≤ sqrt(ϵ) * norm(b)
+
+  LDL1 = spzeros(T, size(A)...)
+  LDL1[diagind(LDL1)] .= one(T)
+  LDL1 = lmul!(S, LDL1)
+  r2 = zeros(T, length(b))
```
Comment (dpo) [inline test/runtests.jl:266]:
```suggestion
  r2 = similar(b)
```

File: test/runtests.jl [hunk 4]
```diff
@@ -230,3 +230,49 @@ end
...
+  mul!(r2, LDL1, x)
+  @test norm(r2 - A * x) ≤ sqrt(ϵ) * norm(b)
```
Comment (dpo) [inline test/runtests.jl:268]:
```suggestion
  @test norm(r2 - A * x) ≤ sqrt(ϵ) * norm(r2)
```

File: test/runtests.jl [hunk 5]
```diff
@@ -230,3 +230,49 @@ end
...
+  r2 .-= b # r2 = LDL*I*x-b
+  @test norm(r2) ≤ sqrt(ϵ) * norm(b)
+  LDL2 = spzeros(T, size(A)...)
+  LDL2[diagind(LDL2)] .= one(T)
+  mul!(LDL2, S, copy(LDL2))
+  mul!(r2, LDL2, x) # same test using mul!
+  @test norm(r2 - A * x) ≤ sqrt(ϵ) * norm(b)
```
Comment (dpo) [inline test/runtests.jl:275]:
```suggestion
  @test norm(r2 - A * x) ≤ sqrt(ϵ) * norm(r2)
```

File: src/LDLFactorizations.jl [hunk 6]
```diff
@@ -329,6 +329,83 @@ function ldl_solve!(n, B::AbstractMatrix{T}, Lp, Li, Lx, D, P) where T
...
+function ldl_mul!(n, x::AbstractVector, Lp, Li, Lx, D, P)
+    @views y = x[P]
+    ldl_ltmul!(n, y, Lp, Li, Lx)
+    ldl_dmul!(n, y, D)
+    ldl_lmul!(n, y, Lp, Li, Lx)
+  return x
+end
+
+# compute L*D*L'*X where X is a matrix
+function ldl_ltmul!(n, X::AbstractMatrix{T}, Lp, Li, Lx) where T
```
Comment (dpo) [inline src/LDLFactorizations.jl:370]: Is it necessary to parametrize here?

**Review by dpo — CHANGES_REQUESTED** [review]

File: src/LDLFactorizations.jl (hunk 6 above)
Comment (geoffroyleconte) [reply src/LDLFactorizations.jl:370]: No it is not. Should I also remove parametrization for ldl_solve!() (lign 324) and for ldl_lsolve!(), ldl_dsolve!() and ldl_ltsolve!() ?

Comment (dpo) [reply src/LDLFactorizations.jl:370]: Sure, but maybe in a separate PR. Thank you!

**Comment by dpo** [comment]
Thank you!

**Pull request merged**